### PR TITLE
[bugfix] Adding protection for empty trajectory

### PR DESCRIPTION
**Pull request by swagata87**
#### PR description:
This is a follow up of my earlier PR on eta extended electrons, which recently got merged  https://github.com/cms-sw/cmssw/pull/35309

It created some issues as reported here https://github.com/cms-sw/cmssw/issues/35488

The issues happened because of missing protection against empty trajectory before accessing trajectory's `lastMeasurement()`. This is being addressed in this PR.

#### PR validation:
Failing WFs reported in github issue https://github.com/cms-sw/cmssw/issues/35488 seem to run fine now, after the fix.

This PR is not a backport.
Backport not necessary.

### CODE REVIEW

**Comment by mmusich** [comment]
type bug-fix

**Comment by mmusich** [comment]
test parameters:
   * workflows = 11630.0, 11650.911, 11723.17, 13034.0, 13034.99

**Comment by mmusich** [comment]
@cmsbuild, please test

File: TrackingTools/TrajectoryFiltering/interface/MinHitsTrajectoryFilter.h [hunk 1]
```diff
@@ -49,16 +49,19 @@ class MinHitsTrajectoryFilter final : public TrajectoryFilter {
...
-    double myEtaSwitch = sinh(theHighEtaSwitch);
-    if (sinhTrajEta2 < (myEtaSwitch * myEtaSwitch)) {
-      if (traj.foundHits() >= theMinHits + seedPenalty)
-        passed = true;
-    } else {  //absTrajEta>theHighEtaSwitch, so apply relaxed cuts
-      if (traj.foundHits() >= theMinHitsAtHighEta + seedPenalty)
-        passed = true;
+
+    if (!traj.empty()) {
+      double pt = traj.lastMeasurement().updatedState().freeTrajectoryState()->momentum().perp();
```
Comment (VinInn) [inline TrackingTools/TrajectoryFiltering/interface/MinHitsTrajectoryFilter.h:54]: why double?  use auto

File: TrackingTools/TrajectoryFiltering/interface/MinHitsTrajectoryFilter.h [hunk 2]
```diff
@@ -49,16 +49,19 @@ class MinHitsTrajectoryFilter final : public TrajectoryFilter {
...
+      double pz = traj.lastMeasurement().updatedState().freeTrajectoryState()->momentum().z();
+      double sinhTrajEta2 = (pz * pz) / (pt * pt);
+      double myEtaSwitch = sinh(theHighEtaSwitch);
```
Comment (VinInn) [inline TrackingTools/TrajectoryFiltering/interface/MinHitsTrajectoryFilter.h:57]: this should go in the constructor

**Comment by qliphy** [comment]
urgent
to fix the IB issue
@swagata87  Please address above comments.

**Comment by swagata87** [comment]
> urgent to fix the IB issue @swagata87 Please address above comments.

the 2nd comment is not clear to me.
`myEtaSwitch ` should go in constructer? 

this is what is there in constructer
```
explicit MinHitsTrajectoryFilter(int minHits = 5,
                                   double highEtaSwitch = 5.0,
                                   int minHitsAtHighEta = 5,
                                   int seedPairPenalty = 0)
      : theMinHits(minHits),
        theHighEtaSwitch(highEtaSwitch),
        theMinHitsAtHighEta(minHitsAtHighEta),
        theSeedPairPenalty(seedPairPenalty) {}
```
`highEtaSwitch` is in constructor and it is in eta. Later we convert it to sinh(eta) just before applying the cut.
` auto myEtaSwitch = sinh(theHighEtaSwitch); `

**Comment by missirol** [comment]
(unrequested comment, since it's 'urgent')

@swagata87 I think the comment is about calculating "sinh2(eta-switch)" inside the constructor (i.e. once), instead of inside the class method (i.e. once per trajectory). Something like

```cpp
  MinHitsTrajectoryFilter(const edm::ParameterSet& pset, edm::ConsumesCollector& iC)
      : theMinHits(pset.getParameter<int>("minimumNumberOfHits")),
        theHighEtaSwitchSinh2(sinh(pset.getParameter<double>("highEtaSwitch"))),
        theMinHitsAtHighEta(pset.getParameter<int>("minHitsAtHighEta")),
        theSeedPairPenalty(pset.getParameter<int>("seedPairPenalty")) { theHighEtaSwitchSinh2 *= theHighEtaSwitchSinh2; }
```

File: TrackingTools/TrajectoryFiltering/interface/MinHitsTrajectoryFilter.h [hunk 3]
```diff
@@ -49,16 +49,19 @@ class MinHitsTrajectoryFilter final : public TrajectoryFilter {
...
-    double myEtaSwitch = sinh(theHighEtaSwitch);
-    if (sinhTrajEta2 < (myEtaSwitch * myEtaSwitch)) {
-      if (traj.foundHits() >= theMinHits + seedPenalty)
-        passed = true;
-    } else {  //absTrajEta>theHighEtaSwitch, so apply relaxed cuts
-      if (traj.foundHits() >= theMinHitsAtHighEta + seedPenalty)
-        passed = true;
+
+    if (!traj.empty()) {
+      auto pt = traj.lastMeasurement().updatedState().freeTrajectoryState()->momentum().perp();
```
Comment (VinInn) [inline TrackingTools/TrajectoryFiltering/interface/MinHitsTrajectoryFilter.h:54]: this should be 
auto pt2 = traj.lastMeasurement().updatedState().freeTrajectoryState()->momentum().perp2();
and later of course auto 
sinhTrajEta2 = (pz * pz) /pt2;

File: TrackingTools/TrajectoryFiltering/interface/MinHitsTrajectoryFilter.h [hunk 4]
```diff
@@ -49,16 +49,19 @@ class MinHitsTrajectoryFilter final : public TrajectoryFilter {
...
+      auto pz = traj.lastMeasurement().updatedState().freeTrajectoryState()->momentum().z();
+      auto sinhTrajEta2 = (pz * pz) / (pt * pt);
+      auto myEtaSwitch = sinh(theHighEtaSwitch);
```
Comment (VinInn) [inline TrackingTools/TrajectoryFiltering/interface/MinHitsTrajectoryFilter.h:57]: this is expensive if done here.
Please move in the constructor IN PLACE of ```theHighEtaSwitch```

**Comment by jpata** [comment]
enable profiling

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
urgent

**Comment by VinInn** [comment]
this is very urgent: hlt is broken.

**Comment by jpata** [comment]
+reconstruction
- bugfix for #35488, introduced here  #35309
- with this PR, the workflows 11630.0, 11650.911, 11723.17, 13034.0, 13034.99 did not crash in step3
- no reco differences expected/observed
- I didn't notice any CPU significant increase in the profiling outputs

**Comment by qliphy** [comment]
+1

**Pull request merged**